### PR TITLE
Port sicmutils.simplify to cljs

### PR DESCRIPTION
**Pull request by sritchie**
Amazing, it works!

### CODE REVIEW

File: src/sicmutils/function.clj [hunk 1]
```diff
@@ -168,22 +169,21 @@
                 (with-meta h {:arity f-arity :from :function-binop}))))]
     (with-meta h {:arity [:exactly 2]})))
 
-(defmacro ^:private make-binary-operation
+(defn ^:private make-binary-operation
```
Comment (sritchie) [inline src/sicmutils/function.clj:172]: tidying as I go. This can be a function.

File: src/sicmutils/generic.cljc [hunk 2]
```diff
@@ -122,6 +124,13 @@
 (defmulti partial-derivative v/argument-kind)
 
 (defmulti simplify v/argument-kind)
+(defmethod simplify :default [a] (v/freeze a))
```
Comment (sritchie) [inline src/sicmutils/generic.cljc:127]: @littleredcomputer , what do you think of putting these default implementations here?

File: src/sicmutils/simplify.cljc [hunk 3]
```diff
@@ -259,30 +267,21 @@
...
 
 (defn expression->string
   "Renders an expression through the simplifier and into a string,
   which is returned."
   [expr]
-  (let [w (StringWriter.)]
-    (expression->stream expr w)
-    (.toString w)))
-(def print-expression #(-> % g/simplify pp/pprint))
+  (with-out-str
```
Comment (sritchie) [inline src/sicmutils/simplify.cljc:283]: cljs compatible, and does the same thing as the previous implementation.

File: test/sicmutils/function_test.clj [hunk 4]
```diff
@@ -20,12 +20,12 @@
 (ns sicmutils.function-test
   (:require [clojure.test :refer :all]
             [sicmutils.generic :as g]
-            [sicmutils.numbers]
+            [sicmutils.matrix :as m]
```
Comment (sritchie) [inline test/sicmutils/function_test.clj:23]: I had to move some functions over, so started this cleanup.

File: test/sicmutils/mechanics/lagrange_test.clj [hunk 5]
```diff
@@ -18,116 +18,137 @@
 ;
 
 (ns sicmutils.mechanics.lagrange-test
-  (:refer-clojure :exclude [+ - * / ref partial zero?])
   (:require [clojure.test :refer :all]
-            [sicmutils.env :refer :all]
+            [sicmutils.calculus.derivative :refer [D]]
```
Comment (sritchie) [inline test/sicmutils/mechanics/lagrange_test.clj:22]: had to move functions over, so removed the `env` import, in the spirit of making the tests work without `env` and with prefixed imports, so we know where everything lives.

File: test/sicmutils/simplify_test.cljc [hunk 6]
```diff
@@ -70,33 +68,19 @@
 (deftest divide-numbers-through
   (is (= 'x (simplify-expression '(* 1 x))))
   (is (= '(* x y z) (simplify-expression '(* 1 x y z))))
-  (is (= 2/3 (simplify-expression '(/ 2 3))))
-  (is (= '(+ (* 1/2 x) (* 1/2 y)) (simplify-expression '(/ (+ x y) 2))))
-  (is (= '(+ x y) (simplify-expression '(/ (* 2 (+ x y)) 2)))))
-
-(deftest equations
```
Comment (sritchie) [inline test/sicmutils/simplify_test.cljc:77]: anything with `literal-function` moved to `sicmutils.function`.

File: test/sicmutils/simplify_test.cljc [hunk 7]
```diff
@@ -114,18 +98,6 @@
   (let [X '(- (expt (sin a) 2) (* (expt (cos b) 2) (expt (sin a) 2)))]
     (is (= '(* (expt (sin a) 2) (expt (sin b) 2)) (simplify-expression X)))))
 
-(deftest lagrange-equations-test
```
Comment (sritchie) [inline test/sicmutils/simplify_test.cljc:117]: these moved to the lagrange equations namespace.

File: test/sicmutils/simplify_test.cljc [hunk 8]
```diff
@@ -144,21 +116,21 @@
...
-         (expression->string ((g/+ (g/square g/sin) (g/square g/tan)) 'x))))
   (is (= "nil" (expression->string nil)))
   (is (= "[nil 3 (+ x 2)]" (expression->string [nil 3 (g/+ 2 'x)])))
-  (is (= "(complex 0.0 1.0)" (expression->string (c/complex 0 1)))))
+  (is (= #?(:clj "(complex 0.0 1.0)"
+            :cljs "(complex 0 1)")
+         (expression->string (c/complex 0 1)))))
 
 (deftest more-trig
-  (is (= '(* -1 (expt (sin x) 2)) (g/simplify (g/+ (g/expt (g/cos 'x) 2) -1))))
```
Comment (sritchie) [inline test/sicmutils/simplify_test.cljc:155]: there is a memoize issue going on, where the order of terms in sums and products can change. Obviously the equations are still equal... I had to move this down to the bottom to get the tests passing.

**Comment by sritchie** [comment]
@littleredcomputer the tests pass here, but Travis is still running and doesn't like this yet (old Node version)

File: src/sicmutils/generic.cljc (hunk 2 above)
Comment (littleredcomputer) [reply src/sicmutils/generic.cljc:127]: I have moved them around many times as I have considered the best place for them.

File: test/sicmutils/mechanics/lagrange_test.clj (hunk 5 above)
Comment (littleredcomputer) [reply test/sicmutils/mechanics/lagrange_test.clj:22]: 👍

**Pull request merged**